lib/pos_blockchain.go

```diff
@@ -100,160 +101,165 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
+		return blockNode, nil
 	}
 
-	// We know the utxoView will be valid because we check that all ancestor blocks have been validated.
+	// We expect the utxoView for the parent block to be valid because we check that all ancestor blocks have
+	// been validated.
 	utxoView, err := bc.getUtxoViewAtBlockHash(*block.Header.PrevBlockHash)
 	if err != nil {
-		// Here we want to at least store this block since we don't know if it's invalid. Hitting
-		// an error here should really never happen.
-		blockNode, storeInBlockIndexErr = bc.storeBlockInBlockIndex(block)
-		if storeInBlockIndexErr != nil {
-			return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding block to block index: %v", storeInBlockIndexErr)
+		// Here we want to at least store this block since we don't know if it's invalid.
+		// Hitting an error here should really never happen.
+		blockNode, innerErr := bc.storeBlockInBlockIndex(block)
+		if innerErr != nil {
+			return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", innerErr)
 		}
-		return blockNode, errors.Wrap(err, "processBlockPoS: Problem initializing UtxoView: ")
+		return blockNode, nil
 	}
-	// Connect this block to the UtxoView.
+
+	// Connect this block to the parent block's UtxoView.
 	txHashes := collections.Transform(block.Txns, func(txn *MsgDeSoTxn) *BlockHash {
 		return txn.Hash()
 	})
-	if _, err = utxoView.ConnectBlock(block, txHashes, true, nil, block.Header.Height); err != nil {
-		// If it doesn't connect, we want to mark it as ValidateFailed.
-		blockNode, storeInBlockIndexErr = bc.storeValidateFailedBlockInBlockIndex(block)
-		if storeInBlockIndexErr != nil {
-			return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding validate failed block to block index: %v", storeInBlockIndexErr)
-		}
-		return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem connecting block to UtxoView: ")
-	}
 
-	validatorsByStake, err := utxoView.GetAllSnapshotValidatorSetEntriesByStake()
-	if err != nil {
-		// Here's another case where we want to store the block even if we hit an error.
-		// We're not SURE if it's invalid, so we should store it, but not mark is as ValidateFailed
-		blockNode, storeInBlockIndexErr = bc.storeBlockInBlockIndex(block)
-		if storeInBlockIndexErr != nil {
-			return blockNode, errors.Wrapf(err, "isValidPoSBlock: Problem adding block to block index: %v", storeInBlockIndexErr)
-		}
-		return blockNode, errors.Wrap(err, "isValidPoSBlock: Problem getting validator set: ")
-	}
-	// TODO(diamondhands): I don't think the TODO below makes sense given the changes but take a look.
-	// TODO: If block belongs in the next epoch or it's from an epoch far ahead in the future, we may not be able to validate its QC at all.
-	// the validator set for that epoch may be entirely different.
+	// TODO: If block belongs in the next epoch or it's from an epoch far ahead in the future, we may not be able to
```

Comment:
I think we want to connect the block before we check the QC. This way the utxo view can use the `CurrentEpochEntry` for "this block" to determine if the QC is valid. I think that why DH moved the connect block to happen here.

Reply:
Makes sense! If the UtxoView can't be built then there's no way we'd be able to validate the QC anyway. Good catch!

Reply:
We need to use the previous block height's validator set to validate an incoming block. The order we perform the QC validation and then the UtxoView construction is correct